### ISSUE TITLE
Feature request: Support Tweedie Distribution

### ISSUE THREAD

**Issue by edgBR** [StatMixedML/XGBoostLSS#11]
Dear colleagues,

According to the docs:

![image](https://user-images.githubusercontent.com/14976422/153877751-c7d4c645-b26b-47f8-84fb-936eb3f039be.png)

Gamma distribution is supported but at the end gamma part of the tweedie family.

Alternative packages already implement this as for instance:

https://github.com/CDonnerer/xgboost-distribution/

BR
E

**Comment by StatMixedML** [comment]
> If I understood correctly XGBoostLSS allows us to predict the entire conditional distribution for our regression target. However you need to specify the underliying distribution first.

Yes correct. One needs to specify a distribution in the first place.

> I am dealing with a problem with heavily zero inflated data (in fact any box cox transformation like log does not change at all the underliying distribution). 

Since I haven't seen the data, I can only speculate on this. But the Negative Binomial is a good choice for count data with many 0s. You may want to try it as an alternative. If this is not working (let me know!), I can also add a Zero-Inflated Negative Binomial, since this is a conceptually more easy than implementing the Tweedie (which has been on my list for quite some time).

**Comment by StatMixedML** [comment]
Hi @edgBR,

> the problem is that my data is actually not count data. Is payment data (continous positive data).

Ok I see. The negative binomial might then not be a suitable assumption.

> As FYI the method that I have built is similar to this: https://scikit-lego.netlify.app/meta.html#Zero-Inflated-Regressor and this methodology has improved my results by 5% in comparison with plain xgboost with the tweedie loss.

I am aware of this approach and this reads like a viable approach for your problem. However, I have some problems with this since you need to estimate two models, one for the 0-part and another one for the non-0-part. More elegantly, you can do the same with a single zero-inflated model. This allows then to also have a proper uncertainty estimation. In principle, you can extend any distribution, say, Normal, Gamma, etc. to account for the excess of 0s. In any case, you need a strong feature that accounts for the 0 part.

**Comment by StatMixedML** [comment]
Hi @edgBR ,

thanks for your detailed answer. I was also to suggest using XGBoostLSS expectile estimation. Let me know how this is going.

**Comment by edgBR** [comment]
@StatMixedML 

For the moment I am getting to wide confidence intervals and the most problematic thing is that the lower bounds contains negative values. Is there a way to put a cap in the expectiles?

**Comment by StatMixedML** [comment]
@edgBR I am afraid there is none, except that you need to manually clip to negative values with 0

**Comment by edgBR** [comment]
@StatMixedML understood.

Would it be possible if you could describe what are the implementation challenges of the tweedie distribution? Is the gradient calculation with autograd the problem?

BR
E

**Comment by StatMixedML** [comment]
@edgBR There is conceptually no problem wit the Tweedie. It has been on my list for some time now. It is just that I haven't spend time on looking into it. Do you know of any PyTorch implementation of the Tweedie?

**Comment by edgBR** [comment]
@StatMixedML a couple of weeks ago I found this in one of the discussions in the gluon-ts github:

https://discuss.pytorch.org/t/custom-tweedie-loss-throwing-an-error-in-pytorch/76349/6

Maybe this is helpful?

**Comment by StatMixedML** [comment]
@edgBR I am aware of this, but I need to check if that is the correct density for the Tweedie. Do you know which density XGBoost is using?

**Comment by edgBR** [comment]
Hi @StatMixedML,

Unfortunatelly I am unaware of what XGBoost is using underneath.

The traces in the code point me here:

https://github.com/tonydifranco/xgboost/blob/4f6d4eaac58c2d93671dbd827acc606b7e8fedf9/src/objective/regression_obj.cc

(Line 285 and beyond) but there I can only find the gradient and the hessian.

In any case this seems like a good implementation of all parameters: https://github.com/thequackdaddy/tweedie/blob/master/tweedie/tweedie_dist.py

But using numpy.

**Comment by StatMixedML** [comment]
@edgBR Thanks for sharing. Looks good!

I am currently working on a different project, but you can try try implementing the tweedie with the autograd yourself and open a PR once working.

**Comment by edgBR** [comment]
Hi @StatMixedML 

Could you maybe facilitate a list of changes that need to be implemented in order to incorporate the tweedie?

If I understood correctly we need:

- AutoGrad Gradient Calculation for each of the parameters
- AutoGrad Hessian Calculation for each of the parameters
- The parameters list for the tweedie distribution (in this case mu, sigma and power between 1 and 2 as you already implemented gamma)
- A way of sampling the pdf of the tweedie

Anything else?

**Comment by StatMixedML** [comment]
@edgBR Besides the things you've mentioned, you would also need

- [Parameter Dictionary](https://github.com/StatMixedML/XGBoostLSS/blob/master/xgboostlss/distributions/Gaussian.py#L39)
- [Inverse Parameter Dictionary](https://github.com/StatMixedML/XGBoostLSS/blob/master/xgboostlss/distributions/Gaussian.py#L52)
- [Starting Values](https://github.com/StatMixedML/XGBoostLSS/blob/master/xgboostlss/distributions/Gaussian.py#L66)
- [Function for calculating quantiles from predicted distribution](https://github.com/StatMixedML/XGBoostLSS/blob/master/xgboostlss/distributions/Gaussian.py#L240)

The last function is needed since it is used for drawing samples from the predicted distribution. In fact, if you need quantiles rather than the samples themselves, the quantile function is more efficient.

**Comment by StatMixedML** [comment]
@edgBR Kindly asking for an update. Did you manage to implement the loss?

**Comment by edgBR** [comment]
Hi @StatMixedML 

Unfortunately not and I have decided to take a new path using MAPIE: https://mapie.readthedocs.io/en/latest/examples_regression/index.html 

plus XGBoost with Tweedie loss.

BR
Edgar

**Comment by StatMixedML** [comment]
@edgBR Thanks for your feedback and for the MAPIE, looks interesting indeed.

**Comment by edgBR** [comment]
Hi @StatMixedML 

Please reopen this.

I found an implementation of tweedie estimation in python which is validated with the R one:

https://github.com/thequackdaddy/tweedie/blob/master/tweedie/tweedie_dist.py

Migrating this to autograd might be easier that expected.

BR
E

**Comment by StatMixedML** [comment]
@edgBR Thanks for sharing, need to look into the details of the implmentation though first.

**Comment by edgBR** [comment]
Hi @StatMixedML I have requested my manager to put some time on this.

So most likely yes :)

**Comment by StatMixedML** [comment]
@edgBR If you want to use PyTorch's autograd, you would first need to translate all functions into pytorch function using tensors. Otherwise, pytorch cannot create the computational graph to calculate derivatives.

**Comment by StatMixedML** [comment]
@edgBR Kindly asking for an update on this. Need advise / help?
